### ISSUE TITLE
Scalability problem

### ISSUE THREAD

**Issue by gvallee** [pmix/prrte#183]
Thank you for taking the time to submit an issue!

## Background information

Sorry if opening the ticket goes again the community rules, I am not quite sure I have yet enough information to have a useful ticket.
I am trying to run scalability tests on various OLCF systems at ORNL to cover some of the needs of some of our users. My current test consist of starting N PEs on X nodes where N is the number of cores available on compute nodes times the number of nodes; basically filling up compute nodes and trying to find the upper limit regarding the number of nodes before we start to face problems. At the moment, I am trying to find the value of X where I start to face problems. For every test, I am running ```hostname``` and to validate the test, I count the number of host names in the output. I acknowledge this might be the best test but it captures the needs from a user; I am willing to run other tests to capture scalability problems. I can also share my test.

### What version of the PMIx Reference Server are you using? (e.g., v1.0, v2.1, git master @ hash, etc.)

PRRTE master fd34cfa8ef3c386acad6d3f3cd4a4c6366a3c3a2

##### What version of PMIx are you using? (e.g., v1.2.5, v2.0.3, v2.1.0, git branch name and hash, etc.)

PMIx master 30c51d72c74f0d225cd60aa8e4ce46054e44603d

### Please describe the system on which you are running

* Operating system/version:
````
$ lsb_release -a
LSB Version:	:core-4.1-noarch:core-4.1-ppc64le
Distributor ID:	RedHatEnterpriseServer
Description:	Red Hat Enterprise Linux Server release 7.5 (Maipo)
Release:	7.5
Codename:	Maipo
````
* Computer hardware:
IBM Power8
* Network type:
Mellanox ConnectX-4
-----------------------------

## Details of the problem

I am currently running my tests on Summitdev at ORNL.
My test runs the following loop starting with 32 nodes and 20 PEs per node (one PE per core):
- if the test fails, try with half the number of nodes
- if the test succeeds, try with twice the number of nodes
- the test stops when we fall back to a single node and when we are back at a number of nodes for which we already had a failure
- all PEs run hostname
- a test is considered as successful when the output has the same count of host names as the number of launched PEs.

On Summitdev, I get the following in a very consistent manner:
- 32 nodes -> failure
- 16 nodes -> failure
- 8 nodes -> success
- 16 nodes -> failure

For the last run with 16 nodes, I get the following error (I do not track error messages for all runs at the moment): ```[summitdev-login1:11547] PMIX ERROR: OUT-OF-RESOURCE in file /ccs/home/gvh/scratch/summitdev/prrte/pmix/src/src/server/pmix_server.c at line 1785
User defined signal 2```
Based on this, I am suspecting a problem in the mapper since it should have all the required resources available.

The LSF script to start a job on 16 nodes looks like:
````
#!/bin/bash
# Begin LSF directives
#BSUB -P *****
#BSUB -J dvm_simple
#BSUB -o dvm_simple.out
#BSUB -e dvm_simple.err
#BSUB -W 00:10
#BSUB -nnodes 16
#BSUB -env "all"
# End LSF directives and begin shell commands

./get_list_hosts.pl

T="$(date +%s)"

echo "Starting DVM on 16 nodes..." >> ./dvm_simple_config.log
prte --prefix $PRRTE_DIR --report-uri prrteuri --hostfile ./DVM_HOSTS.txt &
echo "DVM started" >> ./dvm_simple_config.log

echo "Running job with 320 PEs..." >> ./dvm_simple_config.log
prun --prefix $PRRTE_DIR -np 320 hostname
echo "Job succeeded"  >> ./dvm_simple_config.log

echo "Sleeping for 30 seconds to give a chance to all messages to come back from the nodes..."
sleep 30

echo "Terminating DVM..." >> ./dvm_simple_config.log
prun --prefix $PRRTE_DIR -terminate
echo "DVM teminated" >> ./dvm_simple_config.log

T="$(($(date +%s)-T))"
echo "Total job runtime: $T seconds" >> ./dvm_simple_config.log
````
Note that I included a ```sleep 30``` to give a chance at the system to propagate back all the IO since I believe there is no IO flush in PRRTE at the moment.
This is using a PRRTE module that I generate for the system, PRRTE_DIR points at the install directory for PRRTE.

I will try to run the same test on Summit and Titan to see if I face the same limitations (these systems allow a different number of PEs per node).

Please let me know if you need any additional information, I will be happy to run any test to track this scalability problem.

**Comment by rhc54** [comment]
Hmmm...well, I can tell you the source of the problem. What is happening is that we are caching the output from your ```hostname``` app in the PMIx server that hosts ```prun```. The hotel has a limited amount of space in it - it is intended only to cache output until ```prun``` has a chance to register to receive it. In this case, it is running out of space because of the race condition.

The max size of the hotel is 256 rooms - i.e., we can cache 256 output messages before running out of room. I see that this is a hardcoded limit at the moment - will need to change that, obviously, and rethink how we handle these chatty situations. Meantime, try doing "sleep 10; hostname" as your app so that prun gets a chance to register for output before you start generating it.

BTW: your ```sleep 30``` won't do anything. If prun exits prior to getting its IO, we throw the rest of it away. There is supposedly a flush mechanism to avoid that scenario.

**Comment by rhc54** [comment]
I would check both the return code and count the number of output lines. ```prun``` shouldn't be exiting prior to outputting all the forwarded text - if it is, then we need to know that.

**Comment by rhc54** [comment]
Got it - thanks for the alert! I verified the problem is cleaned up, but please confirm when you get a chance.
